gateway-service/src/main/java/org/zowe/apiml/gateway/ribbon/loadbalancer/predicate/AuthenticationBasedPredicate.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * This program and the accompanying materials are made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-v20.html
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Copyright Contributors to the Zowe Project.
+ */
+package org.zowe.apiml.gateway.ribbon.loadbalancer.predicate;
+
+import com.netflix.niws.loadbalancer.DiscoveryEnabledServer;
+import com.netflix.zuul.context.RequestContext;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Value;
+import org.zowe.apiml.gateway.cache.LoadBalancerCache;
+import org.zowe.apiml.gateway.ribbon.loadbalancer.LoadBalancingContext;
+import org.zowe.apiml.gateway.ribbon.loadbalancer.RequestAwarePredicate;
+import org.zowe.apiml.gateway.ribbon.loadbalancer.model.LoadBalancerCacheRecord;
+import org.zowe.apiml.gateway.security.service.HttpAuthenticationService;
+
+import java.util.Calendar;
+import java.util.Date;
+import java.util.Optional;
+
+import static org.springframework.cloud.netflix.zuul.filters.support.FilterConstants.SERVICE_ID_KEY;
+
+/**
+ * Based on the authentication information decide which instance should be used.
+ * If the user is authenticated and already has routing information stored in cache, use the information.
+ * <p>
+ * There is also terrible overhead as this happens for all instance ids.
+ */
+@RequiredArgsConstructor
+public class AuthenticationBasedPredicate extends RequestAwarePredicate {
+    private final HttpAuthenticationService authenticationService;
+    private final LoadBalancerCache cache;
+
+    @Value("${instance.metadata.apiml.lb.cacheRecordExpirationTimeInHours:8}")
+    int expirationTime;
+
+    @Override
+    public boolean apply(LoadBalancingContext context, DiscoveryEnabledServer server) {
+        RequestContext requestContext = context.getRequestContext();
+        String serviceId = (String) requestContext.get(SERVICE_ID_KEY);
```

Comment:
ServiceId is in the LoadBalancingContext as well, please use it. It is also in the context's Environment abstraction because the whole named context in which this bean exists is named by service id.

Reply:
its here: `context.key`

Reply:
Well it actually isn't as it differs to what is stored in the RequestContext and as we need to use the one in RequestContext later in the filter, it's better to use the same here as well.